fasttld/FastTLDExtract.py

```diff
@@ -49,84 +88,83 @@ def _trie_construct(self, exclude_private_suffix, file_path=''):
         """
         tld_trie = {}
         PublicSuffixList, PrivateSuffixList, AllSuffixList = getPublicSuffixList(file_path)
-        if exclude_private_suffix:
-            SuffixList = PublicSuffixList
-        else:
-            SuffixList = AllSuffixList
+        SuffixList = PublicSuffixList if exclude_private_suffix else AllSuffixList
         for suffix in SuffixList:
             if '.' in suffix:
                 sp = suffix.split('.')
                 sp.reverse()
                 self.nested_dict(tld_trie, sp)
             else:
-                tld_trie.update({
-                    suffix: {
-                        '_END': True
-                    }
-                })
+                tld_trie[suffix] = {'_END': True}
         for key, val in tld_trie.items():
-            if len(tld_trie[key]) == 1:
-                if '_END' in val:
-                    tld_trie.update({
-                        key: True
-                    })
+            if len(val) == 1 and '_END' in val:
+                tld_trie[key] = True
         return tld_trie
-    
+
     def __call__(self, *args, **kwargs):
         return self.extract(*args, **kwargs)
 
-    def extract(self, input, subdomain=True, format=False):
+    def extract(self, raw_url, subdomain=True, format=False):
         """
         Extract suffix and subdomain from a Domain.
-        :param input: 
-        :param subdomain: Output options. This option will reduce efficiency. Maybe 10% 
-        :param format: To format input string.
+        :param raw_url:
+        :param subdomain: Output options. This option will reduce efficiency. Maybe 10%
+        :param format: To format raw_url string.
         :return: Tuple(subdomain, domain, suffix, domain_name)
         >>> FastTLDExtract.extract('www.google.com.hk', subdomain=True)
         >>> ('www', 'google', 'com.hk', 'google.com.hk')
+
+        >>> FastTLDExtract.extract('127.0.0.1', subdomain=True)
+        >>> ('', '127.0.0.1', '', '127.0.0.1')
         """
-        ret_subdomain = ''
-        ret_domain = ''
-        ret_suffix = ''
-        ret_domain_name = ''
+        ret_subdomain = ret_domain = ret_suffix = ret_domain_name = ''
         if format:
-            input = self.format(input)
-        labels = input.split('.')
+            raw_url = self.format(raw_url)
```

Comment:
fasttld can now parse full urls that include the scheme (e.g. https:// ftp:// etc.) and anything after the host subcomponent (e.g. :5000/index.html?id=42 etc.)

Reply:
When I first built the fasttld python module, it was used for processing large dns datasets.

URL should be parsed before call `extract`.

I will merge your great work and make some improvements.